### ISSUE TITLE
Pagination class is not picked up in @action decorated view methods

### ISSUE THREAD

**Issue by bharling** [tfranzel/drf-spectacular#190]
**Describe the bug**
The _get_paginator method in AutoSchema doesn't seem to pick up pagination_class overrides in custom view methods on ModelViewSets - we user this a fair amount in order to provide alternative pagination methods for different access patterns on our project. The generated schema for the view doesn't include any of pagination response schema or parameters. I've checked and the pagination class we're using exposes the correct `get_pagination_schema` methods

**To Reproduce**
```
class CommentsViewSet(ModelViewSet):

    queryset = Comment.objects.all()
    serializer_class = CommentSerializer

    @action(
        detail=False,
        pagination_class=CursorPagination
    )
    def cursor(self, request, *args, **kwargs):
        """
        Uses cursor-based pagination to return chunks of a rapidly changing list resource

        This has the tradeoff that it is not possible to index individual pages ( you can only obtain references to the
        'next' and 'previous' pages ) - but the pagination will be consistent even with rapidly changing data
        """
        queryset = self.filter_queryset(self.get_queryset())
        page = self.paginate_queryset(queryset)
        context = self.get_serializer_context()
        serializer_class = self.get_serializer_class()
        if page is None:
            serializer = serializer_class(queryset, many=True, context=context)
            return Response(serializer.data)
        else:
            serializer = serializer_class(page, many=True, context=context)
            return self.get_paginated_response(serializer.data)
```

Worth noting, we include this @action method via a mixin class to various viewsets on the api.

**Expected behavior**
Expect to see the correct schema for pagination in the exported result but get this instead

![image](https://user-images.githubusercontent.com/1139376/98140197-0e26ad00-1ebd-11eb-977e-5707917a7697.png)

**Comment by bharling** [comment]
Actually - looks like the autoschema is ignoring the `detail=False` attribute on the `@action` decorator, it would be handy if it could pick that up. 

If I add the following decorator to the view then the pagination is added to the schema - looks like spectacular treats all `@action` decorated methods as 'retrieve' actions, rather than inspecting the `detail` value to determine if a list action is more appropriate.

```
    @extend_schema(
        responses={200: QandAQuestionSerializer(many=True)}
    )
    @action(
        detail=False,
        pagination_class=CursorPagination
    )
    def cursor(self, request, *args, **kwargs):
        return super().cursor(request, *args, **kwargs)
```

would be handy if the automatic detection could handle that and remove the need to add an explicit response schema for every view

**Comment by tfranzel** [comment]
hi @bharling, that's quite interesting. a couple of thoughts:

- i believe pagination is simply not defined on custom `@action` as the pagination only happens natively in the `ListModelMixin` and nowhere else. 
- i believe that `detail=True/False` only refers to whether the action is routed at `x/{id}/action` or `x/action` and nothing more. 
- the detection whether an endpoint is `list` or `retrieve` is based on a heuristic with `retrieve` as fallback. without explicit decoration, there is simply not enough information available for an `action` to safely determine that. one important info is that we do not inspect method bodies. thus we can't possibly know if you called `get_paginated_response` inside the body. 
- inferring that `action` is a `list` based on the existence of `pagination_class` on the `action` could be a fallacy, because it defaults to what the view class has, if not explicitly specified on the `action`. 

in summary, i think your 2nd example is the best you can do. to infer more would be very likely a coin flip. does that make sense?

**Comment by bharling** [comment]
Thanks yes that does make some sense - I can work around it I think. I would say it may be worth introspecting the `detail=False` atrribute in the action decorator though? 

the DRF docs do say this should be considered a diferentiator for whether a decorated method should be considered a 'list' or 'detail' action

https://www.django-rest-framework.org/api-guide/viewsets/#introspecting-viewset-actions

Is there a possibility of adding a check for that into this method on the AutoSchema? -> https://github.com/tfranzel/drf-spectacular/blob/master/drf_spectacular/openapi.py#L83

happy to attempt a PR if you think it would be ok

**Comment by tfranzel** [comment]
i see what you are getting at. the documentation can be misinterpreted easily here. i'm very certain they mean something different from what you are hearing. i tried addressing that in my second bullet point. 

`@action(detail=True)` => `/api/x/{id}/action`
`@action(detail=False)` => `/api/x/action` 

in both cases the `action` response body may be anything (`list` or `retrieve`). you cannot assume one or the other. `detail` merely specifies where the router inserts that endpoint. 

does that make more sense?

**Comment by bharling** [comment]
Ah ok understood - yes that makes sense, happy to close this in that case as the extend_schema decorator does the job here.

**Comment by tfranzel** [comment]
@bharling but thanks for pointing it out anyway. i would've liked to improve the detection here too, but DRF is simply too dynamic in this case and we cannot infer from what DRF is not providing.

**Comment by Elixir-MeetThoriya** [comment]
@extend_schema(
        summary="Product Inventory History",
        description="Retrieve the complete inventory history for a specific product.",
        responses={200: InventoryItemHistorySerializer(many=True)}
    )
    @action(
        detail=False, methods=['get'],
        url_path="(?P<product_id>[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89ab][0-9a-fA-F]{3}-[0-9a-fA-F]{12})/history",
        url_name='product_inventory_history', permission_classes=[CompanyUserPermission, ],
        pagination_class=CustomLimitOffsetPagination
    )
    def product_inventory_history(self, request, product_id, **kwargs):
    
    
    can't make paginated schema.. 
    (and for kind of info @action not have "pagination_class" parameter )

**Comment by tfranzel** [comment]
@Elixir-MeetThoriya seems to work for us. This covered in the tests https://github.com/tfranzel/drf-spectacular/blob/27ae6f293a6e8731649216d62ba2a5c8d65691e6/tests/test_regressions.py#L1073

Aslo testing your example, it works exactly as expected and is paginated.

`{'200': {'content': {'application/json': {'schema': {'$ref': '#/components/schemas/PaginatedInventoryItemHistoryList'}}}, 'description': ''}}`
